amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.amqp
+
+import java.util.concurrent.atomic.AtomicReference
+
+import com.rabbitmq.client.Connection
+
+import scala.compat.java8.FunctionConverters._
+
+/**
+ * Only for internal implementations
+ */
+sealed trait AmqpConnectionProvider {
+  def settings: AmqpConnectionSettings
+  def get: Connection
+  def release(connection: Connection): Unit
+}
+
+final object DefaultAmqpConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings): DefaultAmqpConnectionProvider =
+    DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings)
+}
+
+final case class DefaultAmqpConnectionProvider(settings: AmqpConnectionSettings) extends AmqpConnectionProvider {
+  override def get = settings.getConnection
+  override def release(connection: Connection) = if (connection.isOpen) connection.close()
+}
+
+final object ReusableAmqpConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings): ReusableAmqpConnectionProvider =
+    ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings)
+
+  /**
+   * Java API
+   */
+  def create(settings: AmqpConnectionSettings, automaticRelease: Boolean): ReusableAmqpConnectionProvider =
+    ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease)
+}
+
+final case class ReusableAmqpConnectionProvider(settings: AmqpConnectionSettings, automaticRelease: Boolean = true)
+    extends AmqpConnectionProvider {
+  var clients = 0
```

Comment:
I think we discussed this before, this is not threadsafe. Make an internal state ADT like so:

```
private sealed trait State
private case object Empty extends State
privat case object Connecting extends State
private final case class Connected(connection: Connection, clients: Int)
private case object Closing extends State
private val state = new AtomicReference[State](Empty)
```

And do CAS on it for all updates like so: 
```
@tailrec
override def get: Connection = {
  state.get match {
     case Empty =>
        if (state.compareAndSet(Empty, Connecting)) {
          // only one thread will get here, so now we can safely connect
          ... cas swap Connected in ...
       } else get // another thread is connecting, try again
    case Connecting => get // another thread is connecting, try again
    case c @ Connected(connection, clients) => 
       if (state.compareAndSet(c, Connected(connection, clients + 1)) connection
       else get // try again
    ... and so on...
```